ENH: Revise loglike/deviance to correctly handle scale

thequackdaddy opened the pull request:
xref #3773 

I apologize for this taking longer than had hoped, but my busy period was exacerbated by some unforeseen events. But I'm somewhat back now.

This is the first step to solving the above issue. 

Still a lot is missing... I've only really gotten `loglike` to work for all the families except `Gamma`, `Binomial`, `Gaussian`, and `Tweedie` (which has never had `loglike`). I also need to re-work the docstrings. I'm thinking its more logical to have a `loglike_obs` function called in each family and then have a `loglike` function in the `Family` class so that `loglike` will simply be inherited. I think you the doctrings could be elegantly written to handle this too. 

I'm thinking I will work on deviance next and then ciricle back to `loglike`... I feel like R takes some computational shortcuts...

josef-pkt commented:
This pull request introduces 1 alert and fixes 1 - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-196f2b1663f2d320c2eb51b394893659b31e3ca0)

**new alerts:**

* 1 for Unreachable code

**fixed alerts:**

* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
Lunch does marvelous thing. I think all but Gamma and Tweedie are working now.

josef-pkt commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-90cc87ec2651ccdeab474fff71d734e2d80ec040)

**new alerts:**

* 1 for Unreachable code

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
I have most of this working, I think. Deviance and loglike seem good for pretty much everything except loglike for Gaussian. Also, I'm not totally convinced that my binomial loglike is working. 

I think the problem lies in this [special case](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/families/family.py#L610) that I've never really understood (other than to match stata, I think). 

I also need to work on the docstrings... I'll play with that later...

thequackdaddy commented:
OK I realized that with ordinary least squares, its a rather simple adjustment... scale needs to be divided by `nobs` and not `df_resid`. This brings up the problem that this doesn't currently correctly handle when `freq_weights` use classical OLS.

I worked through the docstrings with some drafts.... lotsa changes so there's likely lotsa problems. For `NegativeBinomial`, I realized R (and possibly stata) have different default `scale` behaviors.

I probably need to proofread this. I *believe* this latest commit will pass all unittests. Considering I had to adjust the unittests to make this happen, I'm not sure how valid this truly is. 

I'll probably sleep on this progress.

thequackdaddy commented:
I'd like to re-approach how we do documentation/structure of `family.py`, just to help keep this cleaner. I have an idea I'm borrowing from `scipy`. 

I've come to like the way that `scipy.dists` defines the distributions. There is a basic class e.g., `rv_continuous` with methods like `pdf`, `cdf`., etc. All you have to do is define private functions `_pdf` and `_cdf` and write some documentation at the class level and inherit `rv_continuous` and poof, you've got yourself a new distribution class. 

What I'd propose is that we move all the documentation for each family to the class level and have `Family` define `deviance`, `loglike`, `loglike_obs`, `resid_dev`, and `resid_anscombe`. These methods would call privately defined methods in the subclass. Each individual family would just need `_loglike_obs`, `_resid_dev`, and `_resid_anscombe` defined as well as class-level documentation. If you wanted documentation for the distribution (e.g., the loglikelihood formula), you'd get it from that distribution's class-level documentation (e.g., `Poisson`). If you wanted the documentation for the methods (e.g., `Poisson.loglike`), you'd get it would just be standard documentation from the `Family.loglike` method. 

@josef-pkt  Any thoughts? Maybe do this in another PR? I think it would make maintaining this documentation much easier.

josef-pkt commented:
about documentation/class structure change: 
I'm skeptical (which doesn't mean that I would rule it out at the moment). The main additional feature for the docstrings in scipy is an elaborate machinery to compose and generate the individual docstrings. This is useful for 100 distributions, but I don't think it's worth the overhead for our "few" families where most methods are mainly for internal use.

If the point is mainly to avoid duplication in docstrings, then I think using the template version that we use in several other classes that have elaborate inheritance.

thequackdaddy commented:
@josef-pkt Could you point me to a good example of a template version that statsmodels uses for documentation?

josef-pkt commented:
I'm not sure about good examples because I don't use them much.
Given what I recently saw, discrete Poisson and new GeneralizedPoisson use `%(params)s` to insert the basic params into the docstring. 
We do something similar in several places. It uses basic string substitution, the scipy distribution version uses reversed template, the super class specifies the template and the subclasses specify the distribution string. all subject to: IIRC

josef-pkt commented:
one other thing I remembered:
One of the main purposes of the distinction in scipy.stats.distributions between base class e.g. pdf and distribution class _pdf is that the pdf method does a lot of generic computation so that the distribution specific _pdf method only needs minimal or no input checking (e.g. broadcasting, checking if arguments are in the valid range for the distribution, handle loc and scale) However, this overhead is pretty expensive. Given that our consumers of the family methods are internal, we already have all or most of the input checking done in the model, i.e. GLM.

thequackdaddy commented:
Sounds good. I might approach this in a different PR. My annoyance is that because I've had to re-do the API for a few of the methods, I was trying to make all the copypasta unnecessary. You are right though... the only other family that I guess we could add is multinomial. I'm unaware (generally) of any other family that we'd ever want to add and as this is all used internally, I have difficulty seeing any advantage here. 

Back to this PR, I still need to check a bunch and I'm hoping to find time tomorrow or this weekend at the latest.

thequackdaddy commented:
@josef-pkt Any thoughts on this? 

Basically, my strategery is that within `Family`, `deviance` and `loglike` are defined as the weighted sum of `resid_dev` and `loglike_obs`, respectively. 

I also fixed what I think is a bug with `loglike_obs` and `Gaussian` because the scale used in the result `llf` (and dependents such as `aic`) needs to be divided by `wnobs`, but only for `llf` and dependents.

I gave the docstrings a read and they look pretty good to me. I fixed a few other things for readability-sake as well. 

Can you let me know if this is on a good path?

josef-pkt commented:
This pull request fixes 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-529d7b2cab38aa5f7dad4aa419d5d610bef41eef)

**fixed alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -431,8 +433,8 @@ def loglike_mu(self, mu, scale=1.):
         """
-        return self.family.loglike(mu, self.endog, self.exog, self.iweights,
-                                   scale)
+        return self.family.loglike(self.endog, mu, self.var_weights,
```
Do we want have a default scale that's not 1? Additioanlly, should we use `sum(freq_weights)` as the divisor for scale for the OLS special case (identity-Poisson) and `df_resid` in other cases? 

Unifying this with `llf` in `GLMResults` would be nice.

thequackdaddy commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -442,7 +444,8 @@ def loglike(self, params, scale=None):
         if scale is None:
             scale = self.estimate_scale(expval)
-        llf = self.family.loglike(self.endog, expval, self.iweights, scale)
+        llf = self.family.loglike(self.endog, expval, self.var_weights,
```
Like above should I unify this with `GLMResults`'s ``llf`?

josef-pkt commented:
This pull request fixes 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-1358bfadf469b8ba6fbc9c4d9dffb2425b236d86)

**fixed alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request fixes 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-44443b68ec61de71c4eb13981833ca242bf71d6b)

**fixed alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request fixes 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-7c7e727af132fd7215ad1b385d8daa0b6b84913e)

**fixed alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request fixes 2 alerts - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-9c477f4f57e3bc3d0b945d4dde869251d0f4f555)

**fixed alerts:**

* 1 for Variable defined multiple times
* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
I was roughly reading through the changes. There are too many so I didn't figure out the details yet.
Overall it looks fine and I didn't see any problems. I don't know yet about some of the details or open issues.

Overall unit test coverage is good in terms of lines covered with only a few lines missing. However, more important in this case is what results are now verified and covered for different cases. In the unit tests we still have several cases where we skip specific tests for specific models. 
Are those skips up to date?
e.g. does the comparison glm-gaussian-var_weights agree now with WLS?

As aside, based on the comments and adjustments in the unit tests it sounds like a mess where R and Stata use different definitions of, for example, llf and aic across families.

Is it possible to collect a table to summarize what is verified, e.g. with weight types and families in rows and results in columns (bse-nonrobust, llf, aic, bse-hc0, ...)?

(I'm sorry about the slow response here. Next week the kids go back to school and I should have some quite time to think through the changes. I would also be happy if I don't need to do that and can just merge. :)

thequackdaddy commented:
Sure. I'm having a little bit of rain this weekend so this may help preoccupy my time.

josef-pkt commented:
(I thought Harvey is a giant friendly rabbit.)

thequackdaddy commented:
I finally got around to getting back to making this table. My apologies for the delay. 

I made this in Excel and it could have some problems. Let me know if this is sufficient.

thequackdaddy commented:
```
================================ ====================== ======  ===================== === ======= ======== ============== ============= ============== ============= ============== ====
Test                             Compared To            params  normalized_cov_params bse loglike deviance resid_response resid_pearson resid_deviance resid_working resid_anscombe chi2
================================ ====================== ======  ===================== === ======= ======== ============== ============= ============== ============= ============== ====
TestGlmPoissonPlain              stata                  X                             X   X       X        X              X             X              X             X              X
TestGlmPoissonFwNr               stata                  X                             X   X       X        X              X             X              X             X              X
TestGlmPoissonAwNr               stata                  X                             X   X       X        X              X             X              X             X              X
TestGlmPoissonFwHC               stata                  X                             X   X       X                                                                                 X
TestGlmPoissonAwHC               stata                  X                             X   X       X                                                                                 X
TestGlmPoissonFwClu              stata                  X                             X   X       X                                                                                 X
TestGlmTweedieAwNr               R                      X                             X           X        X              X             X              X
TestGlmGammaAwNr                 R                      X                             X   special X        X              X             X              X
TestGlmGaussianAwNr              R                      X                             X   special X        X              X             X              X
TestRepeatedvsAggregated         statsmodels.GLM        X       X
TestRepeatedvsAverage            statsmodels.GLM        X       X
TestTweedieRepeatedvsAggregated  statsmodels.GLM        X       X
TestTweedieRepeatedvsAverage     statsmodels.GLM        X       X
TestBinomial0RepeatedvsAverage   statsmodels.GLM        X       X
TestBinomial0RepeatedvsDuplicatedstatsmodels.GLM        X       X
TestBinomialVsVarWeights         statsmodels.GLM        X       X                     X
TestGlmGaussianWLS               statsmodels.WLS        X       X                     X
================================ ====================== ======  ===================== === ======= ======== ============== ============= ============== ============= ============== ====
```

josef-pkt commented:
Very good, bse looks pretty good, and it looks good overall.
We might put the table somewhere in the release notes or accessible from the release notes.
Part of it needs explicit documentation (if it's not documented already) that we don't impose equivalence in scale/bse or similar, e.g. Binomial0RepeatedvsAverage or places where we assume QMLE instead of MLE,  IIRC.

I will go back to this during the next few days, it's one of a few things that I definitely want to get into 0.9.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-8e0ca16e5f2d646f206e12ced166edf3e8044223)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
needs a rebase, 
I tried, there are repeated merge conflict in the tests where master can be kept, and one conflicting pep-8 line length cleanup in genereralized_linear_model.

thequackdaddy commented:
I thought I did rebase it?

thequackdaddy commented:
This is what I see it `git gui`

<img width="772" alt="screen shot 2017-10-31 at 7 28 49 pm" src="https://user-images.githubusercontent.com/8186432/32254871-d57e7aae-be71-11e7-9858-6c9037ffe0d3.png">

thequackdaddy commented:
Scratch that... this is what I see. 

I believe the #4098 merge is the head of master, no? 

<img width="750" alt="screen shot 2017-10-31 at 7 33 22 pm" src="https://user-images.githubusercontent.com/8186432/32254966-81a8be70-be72-11e7-96db-f31fd5d24fad.png">

josef-pkt commented:
> I thought I did rebase it?

Yes you did. I had thought so from the github interface. But when I checked out the PR branch, then it was old and without rebase. Most likely I didn't fetch the repo again after your rebase.
(That explains some divergence from what I saw locally and what I saw on github)

My new checkout also has `Merge pull request #4098 from bashtage/var-model-doc-fixes` as parent for the branch, so the rebase is good.

thequackdaddy commented:
> Part of it needs explicit documentation (if it's not documented already) that we don't impose equivalence in scale/bse or similar, e.g. Binomial0RepeatedvsAverage or places where we assume QMLE instead of MLE, IIRC.

The documentation currently has the following disclaimer. Is this sufficient. 

```
    WARNING: Loglikelihood, llf, deviance, etc. is not valid in models where
    scale is equal to 1 (i.e., ``Binomial``, ``NegativeBinomial``, and
    ``Poisson``). If variance weights are specified, then results such as
    ``loglike`` and ``deviance`` are based on a quasi-likelihood
    interpretation. The loglikelihood is not correctly specified in this case,
    and statistics based on it, such AIC or likelihood ratio tests, are not\
    appropriate.
```

Side note: not sure why I wrote loglikelihood and llf. That's obviously the same thing. There's an extra `\` in the last sentence as well I likely fat fingered.

thequackdaddy commented:
I realized I could test the optimizers work too... It doesn't seem to match for the robust co-variance or the `TestBinomial0RepeatedvsAverage` case. Haven't thought that through. 

```
================================= ====================== ====== ===================== === ======= ======== ============== ============= ============== ============= ============== ==== =========
Test                              Compared To            params normalized_cov_params bse loglike deviance resid_response resid_pearson resid_deviance resid_working resid_anscombe chi2 optimizer
================================= ====================== ====== ===================== === ======= ======== ============== ============= ============== ============= ============== ==== =========
TestGlmPoissonPlain               stata                  X                            X   X       X        X              X             X              X             X              X    X
TestGlmPoissonFwNr                stata                  X                            X   X       X        X              X             X              X             X              X    X
TestGlmPoissonAwNr                stata                  X                            X   X       X        X              X             X              X             X              X    X
TestGlmPoissonFwHC                stata                  X                            X   X       X                                                                                 X
TestGlmPoissonAwHC                stata                  X                            X   X       X                                                                                 X
TestGlmPoissonFwClu               stata                  X                            X   X       X                                                                                 X
TestGlmTweedieAwNr                R                      X                            X           X        X              X             X              X                                 X
TestGlmGammaAwNr                  R                      X                            X   special X        X              X             X              X                                 X
TestGlmGaussianAwNr               R                      X                            X   special X        X              X             X              X                                 X
TestRepeatedvsAggregated          statsmodels.GLM        X      X                                                                                                                        X
TestRepeatedvsAverage             statsmodels.GLM        X      X                                                                                                                        X
TestTweedieRepeatedvsAggregated   statsmodels.GLM        X      X                                                                                                                        X
TestTweedieRepeatedvsAverage      statsmodels.GLM        X      X                                                                                                                        X
TestBinomial0RepeatedvsAverage    statsmodels.GLM        X      X
TestBinomial0RepeatedvsDuplicated statsmodels.GLM        X      X                                                                                                                        X
TestBinomialVsVarWeights          statsmodels.GLM        X      X                     X                                                                                                  X
TestGlmGaussianWLS                statsmodels.WLS        X      X                     X                                                                                                  X
================================= ====================== ====== ===================== === ======= ======== ============== ============= ============== ============= ============== ==== =========
```

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-4c3f0e25a61d9537e3ec455b4b89b439fc87c3b0)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
Hmm... 

One of the 2 python 2.7 cases fails under mysterious circumstances. Its close, but off. I tried to create a similar environment on my win32 computer, but was unable to reproduce the error. 

```
(smtest27) C:\Users\User\git\statsmodels>python -c "import statsmodels.api as sm; sm.show_versions()"

INSTALLED VERSIONS
------------------
Python: 2.7.13.final.0

Statsmodels
===========

Installed: 0.8.0.dev0+b082977 (statsmodels)

Required Dependencies
=====================

cython: 0.26 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\Cython)
numpy: 1.11.3 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\numpy)
scipy: 0.19.0 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\scipy)
pandas: 0.20.1 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\pandas)
    dateutil: 2.6.1 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\dateutil)
patsy: 0.4.1 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\patsy)

Optional Dependencies
=====================

matplotlib: 1.5.3 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\matplotlib)
    backend: Qt5Agg
cvxopt: Not installed
joblib: Not installed

Developer Tools
================

IPython: 5.3.0 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\IPython)
    jinja2: Not installed
sphinx: Not installed
    pygments: 2.2.0 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\pygments)
nose: 1.3.7 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages\nose)
pytest: 3.2.1 (C:\ProgramData\Anaconda3\envs\smtest27\lib\site-packages)
virtualenv: Not installed
```

thequackdaddy commented:
I also realized that python 2.7 test passed while a python 3.4 test failed on my personal travis-ci. 

https://travis-ci.org/thequackdaddy/statsmodels/jobs/298168322

I think this means this particular test is not stable. Honestly, we already test that bfgs and IRLS match in another place. Perhaps I can drop this. I don't think its really that meaningful.

thequackdaddy commented:
@josef-pkt I think I am at a stopping point on this. When you get the time to review, let me know.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-80f5e89a889f2d686f9ef0460fb65da29f280126)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented on statsmodels/genmod/tests/test_glm_weights.py:
```diff
@@ -16,6 +40,7 @@
 from statsmodels.genmod.generalized_linear_model import GLM
 from statsmodels.tools.tools import add_constant
 from statsmodels.discrete import discrete_model as discrete
+import pytest
```
pytest imported a second time, is already on line 37

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-fd800f618ad6ea97623f87b280373f9dc0a53b43)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-c85bd7d072f92a9f7e0c1696536045a4c15e3e50)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-65c36c64f9c1ad2935239fd655896a3a2285b160)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
Hello @josef-pkt 

Let me know if you have time to work with me/review this over the next week or so. I should be mainly free and will have time to work on this if you have ideas/things you want me to look at. I'd love to get this merged and move on to more topics.

josef-pkt commented:
@thequackdaddy  I need to look at this before I can say anything. My main guess is that we merge this essentially as is.

Your PR's for Tweedie enhancements also need updating. I only have a rough overview but it would be good if we could get those also merged.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-04323079201bad6fff0a57ba4738eae6d0b3ddaa)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
Thanks, @josef-pkt 

FYI... I think this PR actually has a merge conflict with some of the Tweedie improvements. `var_weights` and `scale` have lots of overlap that I did not fully understand until I got this piece working. 

So let me know once you review. Looking forward to it.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-e122bbfb6456e4ccf609d02716760aba41ac88fe)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
I'm starting to run some notebook examples again

minor issue: 
scale in summary should not use full precision (e.g. gaussian versus WLS test case)
`Scale:                   83.9909580718`

thequackdaddy commented:
So I understand, you are referring to the print formatting for scale? Seems easy enough to fix.

josef-pkt commented:
Yes, looking at the summary the many digits in scale don't look **nice**.  I usually use Poisson or Binomial as GLM examples and never saw this (or paid attention to it)
BTW: comparison GLM-Gaussian and WLS looks good

one small difference in definition, but I think not relevant for changing
In the test case WLS has df_model 2 while glm has df_model = 1, but df_resid agree, the latter is used for inference and more important. (In general, I don't like the constant definition in df_model )

josef-pkt commented:
It looks like GLM-Gaussing with var_weights agrees with WLS for all cov_types
https://gist.github.com/josef-pkt/88ad009fe7e41e0bb09e4777fc383a11

This most likely means that var_weights are generically handled correctly in the sandwich code. Both score_factor and hessian_factor include the iweights, and then hessian and score_obs also do.

josef-pkt commented:
I went through the Poisson notebook in master again, and it looks still the same. Some explanations might not be very clear but looks ok.
Given the explanation in the notebook we still have the open question about whether var_weights are for correctly specified MLE or for Quasi-MLE, i.e. interpretation of llf and similar in case when theoretical scale=1. 
We still keep that open.

josef-pkt commented:
@thequackdaddy  About the merge conflict with Tweedie PR that you mentioned above:
Are there changes that need to be made to the code in this PR or is it all Tweedie specific that can be fixed in the Tweedie PR during or after rebase?

I expect that we can merge this PR soon. I will still look at it today and tomorrow, as time (family) permits.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-9f311ab46f14cbc186922fae06768ef4b80666a5)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
@josef-pkt - Actually, I think its benign. I think there is a PEP8 change that caused a merge conflict for me when I tried... I thought it was more serious than that. 

However, the this PR creates a new function `_estimate_x2_scale` which could be incorporated in the Tweedie PR to simplify the code.

josef-pkt commented:
another extra:
GLM summary doesn't show the cov_type. search doesn't find an issue directly for that
linear_model uses
```
        if hasattr(self, 'cov_type'):
            top_left.append(('Covariance Type:', [self.cov_type]))
```

josef-pkt commented:
checking more features:
fit_constrained: the parameter constraint looks correct, but bse looks wrong when comparing to using offset instead of a single parameter constraint.
I only tried an example and didn't look at the generic fit_constrained code.

We might want to add explicit warnings (warnings.warn) in both fit_regularized and fit_constrained that those are not supported yet with weights and results will be at least partially incorrect.

another check the constrained auxiliary model looks correct and the same as the offset version
```
print(res_glm_c.results_constrained.summary())
res_glm_c.results_constrained.model.var_weights
```
example at end of updated gist (version 2) https://gist.github.com/josef-pkt/88ad009fe7e41e0bb09e4777fc383a11

josef-pkt commented:
BUG, bug-wrong !!!
it looks like fit_constrained with GLM gaussian has wrong bse even without weights.
The problem seems to be that we multiply cov_params a second time with scale

reverse engineering results:
res0_c uses fit_constrained
res0_o uses offset to incorporate the constrained

```
In [43]:
1 /np.sqrt(res0_c.scale) * res0_c.bse.values[0], res0_o.bse.values[0]
Out[43]:
(0.055855555750867605, 0.055855555750867612)
```

GLM.fit_regularized assigns cov to normalized_cov_params, which, I guess, still gets the scale adjustment
`res._results.normalized_cov_params = cov`

(This would be a serious bug if the same problem is in 0.8. I will check tomorrow with a clean example.
This is possible because GLM.fit_constrained was written together with discrete fit_constraint only for discrete cases. unit tests in discrete for poisson and Logit/Binomial)

thequackdaddy commented:
@josef-pkt Based on your recent comments, I added the `cov_type` to the summary, formatted the print scale, and incorporated #4195 into this (hope that's alright). I extended #4195's tests to compare var_weights... looks like it works. 

Do you want me to add warnings that `fit_regularized` and `fit_constrained` are (possibly) wrong? 

Otherwise, ready to merge?

thequackdaddy commented:
Forgot to mention 2 points - 

1. I *think* that I needed to edit your line 

```python
cls.res1._results.cov_params_default = cls.res1.normalized_cov_params
```

to be 

```python
cls.res1._results.cov_params_default = cls.res1.normalized_cov_params * res1.scale
```

Is that OK? Only way I could make the unit tests pass without editing `statsmodels.base.model`

2. Splitting up #4195 is easy enough. I can easily revert incorporating that into this branch... I still have my old branch.

josef-pkt commented:
However, maybe the unit tests with GLM-Gaussian that still fail for me might be fixed with this PR. I need to check more (later tonight).

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-2427bdb189c1b310ed5470f241be4fe69b88bc0c)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -1285,7 +1285,12 @@ def fit_constrained(self, constraints, start_params=None, **fit_kwds):
-        res._results.normalized_cov_params = cov
+
+        # Issue 4195
+        mu = self.predict(params)
```
scale is already computed in the auxiliary model,
fix is in my PR

josef-pkt commented on statsmodels/genmod/tests/test_constrained.py:
```diff
@@ -118,4 +119,93 @@ def init(cls):
+        mod.exog_names[:] = ['const', 'x1', 'x2', 'x3', 'x4']
+        cls.res1 = mod.fit_constrained('x1=0.5')
+        cls.res1._results.cov_params_default = (
+            cls.res1.normalized_cov_params * cls.res1.scale)
```
the test that set cov_params_default are redundant after the fix.

These tests with weights are not in  4195 and will need to be kept during merge conflict resolution and rebase.

josef-pkt commented:
additional change:
the warning for cov_type in Results needs an `upper`
cov_type.startswith('HC') -> cov_type.upper().startswith('HC')

we don't standardize those, but allow arbitrary case in cov_type (checked with either lower() or upper()
(I think I got a warning in one of the examples when I didn't pay attention to the HC0/hc0 spelling)

thequackdaddy commented:
OK. Once #4195 is ready to go, I'll, remove my fix and go with yours, but keep the GLM `var_weights` vs WLS `weights` and make the `upper` fix via an interactive re-base.

Thanks a bunch, @josef-pkt . Very enlightening for me.

josef-pkt commented:
If #4195 comes back green, then I will merge it tonight and come back to this PR tomorrow. (And hopefully I don't find another bug that needs fixing.)
I don't know yet if I will take the time to figure out how to check/verify fit_regularized or whether we just warn as precaution.

josef-pkt commented:
as reminder: after this PR, the precision in the unit test for scale in fit_regularized can be increased, based on my initial experiments.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-9fcbc68d9f3bb116688ff1c2af55d3c78eac186f)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
@thequackdaddy  Can you amend the last commit and decrease the rtol to 1e-10 in
assert_allclose(res1.scale, res2.scale, rtol=0.05) -> assert_allclose(res1.scale, res2.scale, rtol=1e-10)

it passes for me also with 1e-15, so the agreement is very high, but 1e-10 is enough in the unit tests tests

I will do some more spot-checking today. (I don't think I'm patient enough to read through everything again, and test coverage looks very good.)
We/I can do a follow-up with smaller, not directly related changes that came up, e.g. more information in summary, mle_settings, ...

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-2f2d2ff77587aa04b626b24bf7041556f344c0d4)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented:
I updated and largely cleaned the gist for glm-gaussian https://gist.github.com/josef-pkt/88ad009fe7e41e0bb09e4777fc383a11 run on top of a checkout of this PR branch.
The main interesting parts are the summary_col tables to compare different versions with var_weights, cov_type and fit_regularized.
All agree now "out-of-the-box"

thequackdaddy commented:
Cool gist, Josef. However, I noticed that I seem to have lost the code that added. Probably I messed up the rebase. Everything else looks to be there. 

```
        if hasattr(self, 'cov_type'):
            top_right.append(('Covariance Type:', [self.cov_type]))
```

I will amend to re-add that. Sorry.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-25cc2404bce70a78624cf23438bfc654fcaa0fd3)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

josef-pkt commented on statsmodels/genmod/generalized_linear_model.py:
```diff
@@ -1678,12 +1700,15 @@ def summary(self, yname=None, xname=None, title=None, alpha=.05):
-        #smry.add_table_2cols(self, gleft=diagn_left, gright=diagn_right,
-        #                  yname=yname, xname=xname,
-        #                  title="")
+        if hasattr(self, 'cov_type'):
```
this is too late and has no effect.  (the summary instance is already filled in line 1696
it should be right after the top_right list on line 1690 before title

thequackdaddy replied:
That's embarrassing.

josef-pkt commented:
I updated my notebook in the gist again.
I also added some comparisons for fit_regularized.
Looks also fine for var_weights, but the parameterization of the penalization weight differs between GLM and OLS/WLS and so are not directly comparable for larger weights.

So no warning about var_weights are needed in fit_regularized.

josef-pkt commented:
This pull request fixes 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/statsmodels/statsmodels/rev/pr-7ee8f359f728cec4858dc3e8af11e2715debf753)

**fixed alerts:**

* 1 for Variable defined multiple times

---

*Comment posted by [lgtm.com](https://lgtm.com)*

thequackdaddy commented:
@josef-pkt Anything else? Feels pretty good to me. I cov-type is working now.

josef-pkt commented:
I was just playing :)
https://gist.github.com/josef-pkt/ae344b04b264ea3db02da9c363b6aa24 (Gaussian pretending to be Poisson and vice-versa.)
We have the unit tests and know this is supposed to work, but wasn't sure whether the scale estimate can be correctly overwritten.
I also wanted to see if the theoretical results can be easily seen, and how flexible our GLM options are.

I haven't seen any problems today, everything looks good in the examples.

josef-pkt commented:
I added Poisson replicating Binomial to the gist.
I think I'm done playing. I will have a brief look and merge after dinner.

josef-pkt commented:
Too many details to review again. So I'm just merging this.

Thanks @thequackdaddy  Those things are a lot of work, and I'm glad I don't have to do it all myself, or for most parts don't have to do it at all myself.

thequackdaddy commented:
Thanks, Josef. I think we can close #3030 and #3773 as well.

thequackdaddy commented:
And possibly #2351

josef-pkt commented:
Thanks @thequackdaddy  I closed two, but leave #3030 open for now (based on original motivation not because of content)

thequackdaddy commented:
Great. Let me know if I can be of any assistance getting a release out. Happy to help make statsmodels great again.

josef-pkt commented:
Thanks, one task is to add the changes and enhancements to the release notes
very incomplete draft that can be edited is in the wiki https://github.com/statsmodels/statsmodels/wiki/Release-Notes-0.9

If the tweedie PR is not dangerous in the sense of having a not low probability of introducing new problems, then it could still be merged. 
I will go back to checking what needs to be done for the release for the rest of the week, and hopefully merging some smaller fixes.
I was just looking at #3612, trying to figure out based on the updated gist whether that bug is fixed.

josef-pkt commented:
about #3030  GLM class does not have loglike_obs yet for the standard interface. only the family has loglike_obs

josef-pkt commented:
one more notebook in a gist (that's where I was missing loglike_obs)
comparison of residuals exposure versus var_weights. 
I'm not sure about resid_anscombe, but it has unit tests against Stata according to the unit test matrix above
https://gist.github.com/josef-pkt/7157b6ab2929fb16617becbea2ffddd6

thequackdaddy commented:
@josef-pkt 

re anscombe residuals... for reasons I don't remember, I did not weight the residuals by `var_weights`.

https://github.com/statsmodels/statsmodels/blob/master/statsmodels/genmod/generalized_linear_model.py#L206

Easy enough to change it.

josef-pkt commented:
yes, I'm pretty sure we need var_weights in resid_anscombe the same way as scale.
I just realized that we don't use the actual scale in resid_anscombe, with fixed scale=1, the resid_anscombe will in general not be standard normal distributed. 
(I was thinking about doing a MonteCarlo for detection of outlier residuals, but didn't have time yet. Whatever the definition, those should show what scaling we need for the different residuals to be approximately standard normal)

all of the anscombe look like they need `* sqrt(var_weights)` corresponding to `scale**0.5`
binomial uses, for whatever reason:
`...* scale ** 3) ** (1 / 6.)`
which also corresponds to scale**0.5

josef-pkt commented:
using qqplot also shows that resid_anscombe without var_weights has the wrong scale/variance.

an updated version in a new gist
with some outliers added in and qqplot  (note ppplot always adjust the scale, so is not useful for checking scale/variance)
https://gist.github.com/josef-pkt/b04b4fb3f42869df7f7335ba47583ee5

josef-pkt commented:
One more update to the latest notebook in gist 
At the end I added a quick MonteCarlo for very simple correctly specified model (no outliers). I.e. a very nice case.
(One local change in code: Poisson resid_dev computation was warning about negative numbers (around -1e-15) in sqrt. I clipped those to zero. I didn't check more on this)

The agreement between frequencies and standard normal probabilities is good, even higher than I expected, given that Poisson itself is not close to normal.

(I just realized I forgot to reset the random.seed before running the var_weights version)

with exposure
```
[ 0.01  0.5   0.1   0.2 ]
[[ 0.010248  0.498492  0.099052  0.200192]
 [ 0.010006  0.49808   0.100956  0.200286]
 [ 0.010384  0.498084  0.101066  0.200322]]
```

with var_weights (resid_anscombe not corrected yet)
```
[ 0.01  0.5   0.1   0.2 ]
[[  1.00360000e-02   4.98246000e-01   9.86600000e-02   2.00728000e-01]  resid_pearson
 [  1.01880000e-02   4.98024000e-01   1.00556000e-01   2.00946000e-01]  resid_deviance
 [  5.40000000e-05   1.38480000e-02   2.20200000e-03   2.99800000e-03]]  resid_anscombe
```
